x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/enrich_signal_threat_matches.test.ts

```diff
@@ -0,0 +1,464 @@
+        },
+      },
+    ]);
+  });
+
+  it('returns data sans atomic from first indicator if indicator field is an array of objects', () => {
+    threats = [
+      getThreatListItemMock({
+        _id: '123',
+        _source: {
+          threat: {
+            indicator: [
+              { domain: 'foo', type: 'first' },
+              { domain: 'bar', type: 'second' },
+            ],
+          },
+        },
+      }),
+    ];
+
+    const indicators = buildMatchedIndicator({
+      queries,
+      threats,
+    });
+
+    expect(indicators).toEqual([
+      {
+        domain: 'foo',
+        matched: {
+          atomic: undefined,
+          field: 'event.field',
+          type: 'first',
+        },
+        type: 'first',
+      },
+    ]);
+  });
+
+  it('throws an error if indicator field is not an object or an array', () => {
+    threats = [
+      getThreatListItemMock({
+        _id: '123',
+        _source: {
+          threat: {
+            indicator: 'not an object',
+          },
+        },
+      }),
+    ];
+
+    expect(() =>
+      buildMatchedIndicator({
+        queries,
+        threats,
+      })
+    ).toThrowError('Expected indicator field to be an object, but found: not an object');
+  });
+});
+
+describe('enrichSignalThreatMatches', () => {
+  let getMatchedThreats: GetMatchedThreats;
+  let matchedQuery: string;
+
+  beforeEach(() => {
+    getMatchedThreats = async () => [
+      getThreatListItemMock({
+        _id: '123',
+        _source: {
+          threat: { indicator: { domain: 'domain_1', other: 'other_1', type: 'type_1' } },
+        },
+      }),
+    ];
+    matchedQuery = encodeThreatMatchNamedQuery(
+      getNamedQueryMock({ id: '123', field: 'event.field', value: 'threat.indicator.domain' })
+    );
+  });
+
+  it('performs no enrichment if there are no signals', async () => {
+    const signals = getSignalsResponseMock([]);
+    const enrichedSignals = await enrichSignalThreatMatches(signals, getMatchedThreats);
+
+    expect(enrichedSignals.hits.hits).toEqual([]);
+  });
+
+  it('preserves existing threat.indicator objects on signals', async () => {
+    const signalHit = getSignalHitMock({
+      _source: { '@timestamp': 'mocked', threat: { indicator: [{ existing: 'indicator' }] } },
+      matched_queries: [matchedQuery],
+    });
+    const signals = getSignalsResponseMock([signalHit]);
+    const enrichedSignals = await enrichSignalThreatMatches(signals, getMatchedThreats);
+    const [enrichedHit] = enrichedSignals.hits.hits;
+    const indicators = get(enrichedHit._source, 'threat.indicator');
+
+    expect(indicators).toEqual([
+      { existing: 'indicator' },
+      {
+        domain: 'domain_1',
+        matched: { atomic: 'domain_1', field: 'event.field', type: 'type_1' },
+        other: 'other_1',
+        type: 'type_1',
+      },
+    ]);
+  });
+
+  // TODO note in PR
+  it('provides only match data if the matched threat cannot be found', async () => {
```

Comment:
I wanted to call this out as a behavioral edge case: if the indicator index is modified during rule execution such that an indicator matched during query time, but was deleted before the subsequent enrichment, this would be the resulting behavior.